src/completion.ts

```diff
@@ -105,22 +107,34 @@ export function asCompletionItem(entry: tsp.CompletionEntry, optionalReplacement
             }
         }
     }
-    if (replacementRange) {
-        if (!insertText) {
-            insertText = item.label;
-        }
-        if (features.completionInsertReplaceSupport) {
-            const insertRange = lsp.Range.create(replacementRange.start, position);
-            item.textEdit = lsp.InsertReplaceEdit.create(insertText, insertRange, replacementRange);
-        } else {
-            item.textEdit = lsp.TextEdit.replace(replacementRange, insertText);
-        }
-    } else {
-        item.insertText = insertText;
-    }
     return item;
 }
 
+function getRangeFromReplacementSpan(
```

Comment:
this function is pretty similar to what VS Code does:

```ts
    // vscode/extensions/typescript-language-features/src/languageFeatures/completions.ts
	private getRangeFromReplacementSpan(tsEntry: Proto.CompletionEntry, completionContext: CompletionContext) {
		if (!tsEntry.replacementSpan) {
			return;
		}

		let replaceRange = typeConverters.Range.fromTextSpan(tsEntry.replacementSpan);
		// Make sure we only replace a single line at most
		if (!replaceRange.isSingleLine) {
			replaceRange = new vscode.Range(replaceRange.start.line, replaceRange.start.character, replaceRange.start.line, completionContext.line.length);
		}

		// If TS returns an explicit replacement range, we should use it for both types of completion
		return {
			inserting: replaceRange,
			replacing: replaceRange,
		};
	}
```

So basically if an entry has a replacementSpan, VS Code uses that replacementSpan for both insert and replace.

You took a little different approach, but it works.

Reply:
VSCode doesn't use LSP so we can't use exactly same code but yes, it's based on VSCode's logic.